File: app/src/net/reichholf/dreamdroid/helpers/PiconSyncService.java
package net.reichholf.dreamdroid.helpers;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.support.v7.preference.PreferenceManager;
import android.util.Log;

import net.reichholf.dreamdroid.DreamDroid;
import net.reichholf.dreamdroid.Profile;
import net.reichholf.dreamdroid.R;
import net.reichholf.dreamdroid.helpers.enigma2.Picon;

import java.io.File;

import it.sauronsoftware.ftp4j.FTPClient;
import it.sauronsoftware.ftp4j.FTPFile;

/**
 * Created by Stephan on 05.02.2016.
 */
public class PiconSyncService extends IntentService {
	private static final String TAG = PiconSyncService.class.getSimpleName();
	int mId = 0x9923;

	public class DownloadProgress {
		public static final int EVENT_ID_CONNECTING = 0;
		public static final int EVENT_ID_CONNECTED = 1;
		public static final int EVENT_ID_LOGIN_SUCCEEDED = 2;
		public static final int EVENT_ID_LISTING = 3;
		public static final int EVENT_ID_LISTING_READY = 4;
		public static final int EVENT_ID_DOWNLOADING_FILE = 6;
		public static final int EVENT_ID_FINISHED = 7;
		public static final int EVENT_ID_ERROR = 0x99;

		public boolean connected;
		public boolean error;
		public int totalFiles;
		public int downloadedFiles;
		public String currentFile;
		public String errorText;

		public DownloadProgress() {
			connected = false;
			error = false;
			totalFiles = 0;
			downloadedFiles = 0;
			currentFile = "";
			errorText = "";
		}
	}

	protected NotificationManagerCompat mNotifyManager;
	protected NotificationCompat.Builder mNotificationBuilder;
	protected DownloadProgress mDownloadProgress;

	/**
	 * Creates an IntentService for syncing picons
	 */
	public PiconSyncService() {
		super(PiconSyncService.class.getCanonicalName());
	}

	@Override
	protected void onHandleIntent(Intent intent) {
		mDownloadProgress = new DownloadProgress();
		initNotifications();
		syncPicons();
	}

	protected void syncPicons() {
		String localPath = Picon.getBasepath(getApplicationContext());
		String remotePath = PreferenceManager.getDefaultSharedPreferences(getBaseContext()).getString(DreamDroid.PREFS_KEY_SYNC_PICONS_PATH, "/usr/share/enigma2/picon");
		Log.i(TAG, String.format("Syncing from %s to %s", remotePath, localPath));
		FTPClient client = new FTPClient();
		Profile p = DreamDroid.getCurrentProfile();
		try {
			// Check for the required directories
			File tmpFile = new File(localPath);
			if (!tmpFile.exists())
				tmpFile.mkdirs();

			tmpFile = new File(String.format("%s.nomedia", localPath));
			if (!tmpFile.exists())
				tmpFile.createNewFile();

			publishProgress(DownloadProgress.EVENT_ID_CONNECTING);
			client.connect(p.getHost());
			publishProgress(DownloadProgress.EVENT_ID_CONNECTED);
			client.login(p.getUser(), p.getPass());
			publishProgress(DownloadProgress.EVENT_ID_LOGIN_SUCCEEDED);

			client.setType(FTPClient.TYPE_BINARY);
			Log.i(TAG, String.format("Changing to %s", remotePath));
			client.changeDirectory(remotePath);
			publishProgress(DownloadProgress.EVENT_ID_LISTING);

			FTPFile[] fileList = client.list("*.png");
			mDownloadProgress.totalFiles = fileList.length;
			publishProgress(DownloadProgress.EVENT_ID_LISTING_READY);
			for (FTPFile remoteFile : fileList) {
				if (remoteFile.getType() != FTPFile.TYPE_FILE)
					continue;
				String fileName = remoteFile.getName();

				mDownloadProgress.currentFile = fileName;
				publishProgress(DownloadProgress.EVENT_ID_DOWNLOADING_FILE);

				File localFile = new File(String.format("%s%s", localPath, fileName));
				localFile.createNewFile();

				try {
					client.download(fileName, localFile);
				} catch (Exception e) {
					e.printStackTrace();
					Log.e(TAG, "Failed to download picon with filename " + fileName);
				}
				mDownloadProgress.downloadedFiles++;
			}

		} catch (Exception e) {
			e.printStackTrace();
			mDownloadProgress.error = true;
			mDownloadProgress.errorText = e.getMessage();
			publishProgress(DownloadProgress.EVENT_ID_ERROR);
		}
		publishProgress(DownloadProgress.EVENT_ID_FINISHED);
	}

	public void initNotifications() {
		Context context = getApplicationContext();
		mNotifyManager = NotificationManagerCompat.from(context);
		mNotificationBuilder = new NotificationCompat.Builder(context);

		Bitmap bm = BitmapFactory.decodeResource(context.getResources(), R.mipmap.ic_launcher);
		mNotificationBuilder.setContentTitle(context.getString(R.string.sync_picons))
				.setLargeIcon(bm)
				.setSmallIcon(R.drawable.ic_action_refresh);
	}

	public void publishProgress(int eventid) {
		String message = "-";
		switch (eventid) {
			case DownloadProgress.EVENT_ID_CONNECTING:
				message = getString(R.string.connecting);
				break;
			case DownloadProgress.EVENT_ID_CONNECTED:
				message = getString(R.string.connected);
				break;
			case DownloadProgress.EVENT_ID_LOGIN_SUCCEEDED:
				message = getString(R.string.connected);
				break;
			case DownloadProgress.EVENT_ID_LISTING:
				message = getString(R.string.getting_list_of_files);
				break;
			case DownloadProgress.EVENT_ID_LISTING_READY:
				message = getString(R.string.checking);
				break;
			case DownloadProgress.EVENT_ID_DOWNLOADING_FILE:
				message = mDownloadProgress.currentFile;
				break;
			case DownloadProgress.EVENT_ID_FINISHED:
				Picon.clearCache();
				if (!mDownloadProgress.error) {
					message = getString(R.string.picon_sync_finished, mDownloadProgress.downloadedFiles);
				} else {
					message = mDownloadProgress.errorText;
					if (message == null) //TODO I am not happy about this, imo this shouldn't even happen!
						message = mDownloadProgress.currentFile;
				}
				break;
		}
		if (mDownloadProgress.totalFiles > 0)
			mNotificationBuilder.setContentText(message).setProgress(mDownloadProgress.totalFiles, mDownloadProgress.downloadedFiles, false).setOngoing(true);
		else
			mNotificationBuilder.setContentText(message).setOngoing(true);
		if (eventid == DownloadProgress.EVENT_ID_FINISHED)
			mNotificationBuilder.setOngoing(false);
		mNotifyManager.notify(mId, mNotificationBuilder.build());

	}
}

File: app/src/net/reichholf/dreamdroid/helpers/PiconSyncService.kt
package net.reichholf.dreamdroid.helpers

import android.app.IntentService
import android.content.Intent
import android.graphics.BitmapFactory
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.preference.PreferenceManager
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import net.reichholf.dreamdroid.DreamDroid
import net.reichholf.dreamdroid.R
import net.reichholf.dreamdroid.helpers.enigma2.Picon
import org.apache.commons.net.ftp.FTPClient
import org.apache.commons.net.ftp.FTPClientConfig
import org.apache.commons.net.ftp.FTPFileFilter

/**
 * Created by Stephan on 05.02.2016.
 */
class PiconSyncService : IntentService(PiconSyncService::class.java.canonicalName) {
    private val id = 0x9923

    class DownloadProgress {
        var connected: Boolean = false
        var error: Boolean = false
        var totalFiles: Int = 0
        var downloadedFiles: Int = 0
        var currentFile: String = ""
        var errorText: String? = ""

        companion object {
            const val EVENT_ID_CONNECTING: Int = 0
            const val EVENT_ID_CONNECTED: Int = 1
            const val EVENT_ID_LOGIN_SUCCEEDED: Int = 2
            const val EVENT_ID_LISTING: Int = 3
            const val EVENT_ID_LISTING_READY: Int = 4
            const val EVENT_ID_DOWNLOADING_FILE: Int = 6
            const val EVENT_ID_FINISHED: Int = 7
            const val EVENT_ID_ERROR: Int = 0x99
        }
    }

    protected lateinit var notifyManager: NotificationManagerCompat
    protected lateinit var notificationBuilder: NotificationCompat.Builder
    protected lateinit var downloadProgress: DownloadProgress

    override fun onHandleIntent(intent: Intent?) {
        downloadProgress = DownloadProgress()
        initNotifications()
        syncPicons()
    }

    protected fun syncPicons() {
        val localPath = Picon.getBasepath(applicationContext)
        val remotePath = PreferenceManager.getDefaultSharedPreferences(baseContext)
            .getString(DreamDroid.PREFS_KEY_SYNC_PICONS_PATH, "/usr/share/enigma2/picon")
        Log.i(TAG, String.format("Syncing from %s to %s", remotePath, localPath))
        val client = FTPClient()

        @Suppress("UNUSED_VARIABLE")
        val config = FTPClientConfig()
        val p = DreamDroid.getCurrentProfile()
        try {
            var tmpFile = File(localPath)
            if (!tmpFile.exists()) {
                tmpFile.mkdirs()
            }

            tmpFile = File(String.format("%s.nomedia", localPath))
            if (!tmpFile.exists()) {
                tmpFile.createNewFile()
            }

            publishProgress(DownloadProgress.EVENT_ID_CONNECTING)
            client.connect(p.host)
            publishProgress(DownloadProgress.EVENT_ID_CONNECTED)
            client.login(p.user, p.pass)
            publishProgress(DownloadProgress.EVENT_ID_LOGIN_SUCCEEDED)

            client.setFileType(FTPClient.BINARY_FILE_TYPE)
            Log.i(TAG, String.format("Changing to %s", remotePath))
            client.changeWorkingDirectory(remotePath)
            publishProgress(DownloadProgress.EVENT_ID_LISTING)

            val filter = FTPFileFilter { file -> file.isFile && file.name.endsWith(".png") }
            val fileList = client.listFiles(null, filter)
            downloadProgress.totalFiles = fileList.size
            publishProgress(DownloadProgress.EVENT_ID_LISTING_READY)
            for (remoteFile in fileList) {
                if (!remoteFile.isFile) continue
                val fileName = remoteFile.name

                downloadProgress.currentFile = fileName
                publishProgress(DownloadProgress.EVENT_ID_DOWNLOADING_FILE)

                val localFile = File(String.format("%s%s", localPath, fileName))
                localFile.createNewFile()
                val outputStream = BufferedOutputStream(FileOutputStream(localFile))
                try {
                    client.retrieveFile(fileName, outputStream)
                } catch (e: Exception) {
                    e.printStackTrace()
                    Log.e(TAG, "Failed to download picon with filename $fileName")
                }
                outputStream.close()
                downloadProgress.downloadedFiles++
            }
        } catch (e: Exception) {
            e.printStackTrace()
            downloadProgress.error = true
            downloadProgress.errorText = e.message
            publishProgress(DownloadProgress.EVENT_ID_ERROR)
        }
        publishProgress(DownloadProgress.EVENT_ID_FINISHED)
    }

    fun initNotifications() {
        val context = applicationContext
        notifyManager = NotificationManagerCompat.from(context)
        notificationBuilder = NotificationCompat.Builder(context, "dreamdroid_picon_sync")

        val bm = BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
        notificationBuilder.setContentTitle(context.getString(R.string.sync_picons))
            .setLargeIcon(bm)
            .setSmallIcon(R.drawable.ic_action_refresh)
    }

    fun publishProgress(eventid: Int) {
        var message = "-"
        when (eventid) {
            DownloadProgress.EVENT_ID_CONNECTING -> message = getString(R.string.connecting)

            DownloadProgress.EVENT_ID_CONNECTED -> message = getString(R.string.connected)

            DownloadProgress.EVENT_ID_LOGIN_SUCCEEDED -> message = getString(R.string.connected)

            DownloadProgress.EVENT_ID_LISTING -> message = getString(R.string.getting_list_of_files)

            DownloadProgress.EVENT_ID_LISTING_READY -> message = getString(R.string.checking)

            DownloadProgress.EVENT_ID_DOWNLOADING_FILE -> message = downloadProgress.currentFile

            DownloadProgress.EVENT_ID_FINISHED -> {
                Picon.clearCache(this)
                if (!downloadProgress.error) {
                    message =
                        getString(R.string.picon_sync_finished, downloadProgress.downloadedFiles)
                } else {
                    message = downloadProgress.errorText ?: downloadProgress.currentFile
                }
            }
        }
        if (downloadProgress.totalFiles > 0) {
            notificationBuilder.setContentText(message)
                .setProgress(downloadProgress.totalFiles, downloadProgress.downloadedFiles, false)
                .setOngoing(true)
        } else {
            notificationBuilder.setContentText(message).setOngoing(true)
        }
        if (eventid == DownloadProgress.EVENT_ID_FINISHED) {
            notificationBuilder.setOngoing(false)
        }
        notifyManager.notify(id, notificationBuilder.build())
    }

    companion object {
        private val TAG = PiconSyncService::class.java.simpleName
    }
}
